code/headers/gcode_generator.hpp

```diff
@@ -0,0 +1,136 @@
+#pragma once
+#include "vector3.hpp"
+#include <hwlib.hpp> // size_t, (u)intx_t
+
+namespace r2d2::robot_arm {
+    /**
+     * Template Interface class for generating gcodes.
+     * Has a few helper functions such as converting an int to a char.
+     * Size represents the size of the char buffer array.
+     * Note that a string of 4 characters, has a size of 5 with the '\0'
+     * character
+     * @param Size buffer size
+     * */
+    template <size_t Size>
+    class gcode_generator_c {
+    protected:
+        /**
+         * Buffer with Size as size.
+         * */
+        char buffer[Size] = {'\0'};
```

Comment:
I don't know how much you change this buffer in child classes, but it might be worthwhile to cache the length of the buffer.